src/bin/inx-chronicle/api/error.rs

```diff
@@ -4,140 +4,205 @@
+pub enum CorruptStateError {
+    #[error("no milestone in the database")]
+    Milestone,
+    #[error("no node configuration in the database")]
+    NodeConfig,
+    #[error("no protocol parameters in the database")]
+    ProtocolParams,
+}
+
+impl ErrorStatus for CorruptStateError {
+    fn status(&self) -> StatusCode {
+        StatusCode::INTERNAL_SERVER_ERROR
+    }
 }
 
 #[derive(Error, Debug)]
 #[allow(missing_docs)]
-pub enum ApiError {
-    #[error(transparent)]
-    BadParse(#[from] ParseError),
-    #[error("Invalid time range")]
-    BadTimeRange,
-    #[error("Invalid password provided")]
+pub enum AuthError {
+    #[error("invalid password provided")]
     IncorrectPassword,
-    #[error("Internal server error")]
-    Internal(InternalApiError),
-    #[error(transparent)]
+    #[error("invalid JWT provided: {0}")]
     InvalidJwt(auth_helper::jwt::Error),
-    #[error(transparent)]
-    InvalidAuthHeader(#[from] TypedHeaderRejection),
-    #[error("No results returned")]
-    NoResults,
-    #[error("No endpoint found")]
-    NotFound,
-    #[error("Endpoint not implemented")]
-    NotImplemented,
-    #[error(transparent)]
-    QueryError(#[from] QueryRejection),
 }
 
-impl ApiError {
-    /// Gets the HTTP status code associated with this error.
-    pub fn status(&self) -> StatusCode {
-        match self {
-            ApiError::NoResults | ApiError::NotFound => StatusCode::NOT_FOUND,
-            ApiError::BadTimeRange
-            | ApiError::BadParse(_)
-            | ApiError::InvalidAuthHeader(_)
-            | ApiError::QueryError(_) => StatusCode::BAD_REQUEST,
-            ApiError::Internal(_) => StatusCode::INTERNAL_SERVER_ERROR,
-            ApiError::IncorrectPassword | ApiError::InvalidJwt(_) => StatusCode::UNAUTHORIZED,
-            ApiError::NotImplemented => StatusCode::NOT_IMPLEMENTED,
-        }
+impl ErrorStatus for AuthError {
+    fn status(&self) -> StatusCode {
+        StatusCode::UNAUTHORIZED
     }
+}
 
-    /// Gets the u16 status code representation associated with this error.
-    pub fn code(&self) -> u16 {
-        self.status().as_u16()
+#[derive(Error, Debug)]
+#[allow(missing_docs)]
+#[error("endpoint not implemented")]
+pub struct UnimplementedError;
+
+impl ErrorStatus for UnimplementedError {
+    fn status(&self) -> StatusCode {
+        StatusCode::NOT_IMPLEMENTED
     }
+}
 
-    /// Creates a new ApiError from a bad parse.
-    pub fn bad_parse(err: impl Into<ParseError>) -> Self {
-        ApiError::BadParse(err.into())
+impl IntoResponse for UnimplementedError {
+    fn into_response(self) -> axum::response::Response {
+        ApiError::from(self).into_response()
     }
 }
 
-impl<T: Into<InternalApiError>> From<T> for ApiError {
-    fn from(err: T) -> Self {
-        ApiError::Internal(err.into())
+#[derive(Error, Debug)]
+#[allow(missing_docs)]
+pub enum MissingError {
+    #[error("no results returned")]
+    NoResults,
+    #[error("no endpoint found")]
+    NotFound,
```

Comment:
I personally think that `NotFound` is not clear enough IMO (you could also say that about data). How about "NoEndpoint", or `EndpointNotFound` to make it unambiguous?

Reply:
Talk to the HTTP protocol creators?

Reply:
Well ... the rebellion has to start somewhere :trollface: